src/treelearner/serial_tree_learner.cpp

```diff
@@ -768,7 +768,7 @@ double SerialTreeLearner::GetParentOutput(const Tree* tree, const LeafSplits* le
   return parent_output;
 }
 
-void SerialTreeLearner::RecomputeBestSplitForLeaf(int leaf, SplitInfo* split) {
+void SerialTreeLearner::RecomputeBestSplitForLeaf(Tree* tree, int leaf, SplitInfo* split) {
```

Comment:
sorry for not reveiewing sooner, but this is a breaking change to the library's API. is that ok?

Just want to be sure we consider it before this is merged

Reply:
I guess it is internal API not public C API, isn't it?

Reply:
that method is marked `protected` (https://github.com/microsoft/LightGBM/blob/fcfd4132e6d40a22d52023396329c41fd3de4a42/src/treelearner/serial_tree_learner.h#L121).

I'm not that experienced with C++, but from https://docs.microsoft.com/en-us/cpp/cpp/protected-cpp?view=msvc-160#remarks it seems `protected` methods aren't exactly "internal".

> Protected members are not as private as private members, which are accessible only to members of the class in which they are declared, but they are not as public as public members, which are accessible in any function.

> Class members declared as protected can be used only by the following....Classes derived with public or protected access from the class that originally declared these members.

Reading that, it seems to me that someone writing their own extension of LightGBM could be relying on this API.

@btrotta could you comment on this? If I'm totally wrong here I apologize for slowing down the PR and would be happy to remove my blocking review.

Reply:
@jameslamb 
> Reading that, it seems to me that someone writing their own extension of LightGBM could be relying on this API.

The only API for which we are trying to provide backward compatibility is C API and all functions from it are listed in this file: https://github.com/microsoft/LightGBM/blob/master/include/LightGBM/c_api.h
All other functions are treated as internal and we do not make any promises regarding their signatures.

Reply:
alright, my review can be dismissed then. Sorry to slow the process down.

Reply:
thanks you nevertheless for your comments.